### DISCUSSION TITLE
Issues with Debian 11 Bullseye / Python 3.9

### DISCUSSION THREAD

**Discussion by ghost** [modoboa/modoboa#2386] [Help]
I updated my server from Debian 10.11 (Buster) with Python 3.7.3 to Debian 11.1 (Bullseye) with Python 3.9.2. Modoboa 1.17.0 (latest) was working on Debian 10.11. Not sure but I think the errors are Python related.

**OS**: Debian 11.1 Bullseye
**Python**: 3.9.2
**Modoboa**: 1.17.0 (latest, modoboa-installer)
**Webmail**: 1.6.1

### AttributeError at /
**Note**: This error also appears on a fresh Debian 11.1 install in VirtualBox. So it's not necessary an upgrade failure.

```
AttributeError at /
module 'base64' has no attribute 'decodestring'

Request Method: GET
Request URL: https://mail.<mydomain>.com/
Django Version: 2.2.24
Exception Type: AttributeError
Exception Value: module 'base64' has no attribute 'decodestring'
Exception Location: /srv/modoboa/env/lib/python3.9/site-packages/feedparser.py in <module>, line 93

Python Executable: /usr/bin/uwsgi-core
Python Version: 3.9.2

Python Path: ['.',
              '',
              '/usr/lib/python39.zip',
              '/usr/lib/python3.9',
              '/usr/lib/python3.9/lib-dynload',
              '/srv/modoboa/env/lib/python3.9/site-packages']
```

### Fix:
**Note**: Don't know if this is a good fix, but seems to work for now.
```
sed -i 's/base64.decodestring/base64.decodebytes/g' /srv/modoboa/env/lib/python3.9/site-packages/feedparser.py
```

### UnboundLocalError at /webmail/
At this point it looks like everything is working except for /webmail/ which is not very convenient :-)

```
UnboundLocalError at /webmail/
local variable 'data' referenced before assignment

Request Method: GET
Request URL: https://mail.<mydomain>.com/webmail/
Django Version: 2.2.24
Exception Type: UnboundLocalError
Exception Value: local variable 'data' referenced before assignment
Exception Location: /srv/modoboa/env/lib/python3.9/site-packages/modoboa_webmail/lib/imaputils.py in getquota, line 691

Python Executable: /usr/bin/uwsgi-core
Python Version: 3.9.2

Python Path: ['.',
              '',
              '/usr/lib/python39.zip',
              '/usr/lib/python3.9',
              '/usr/lib/python3.9/lib-dynload',
              '/srv/modoboa/env/lib/python3.9/site-packages',
              '/usr/local/lib/python3.9/dist-packages',
              '/usr/lib/python3/dist-packages']
```

**Traceback**

```
File "/srv/modoboa/env/lib/python3.9/site-packages/modoboa_webmail/lib/imaputils.py" in getquota
  686.             data = self._cmd("GETQUOTAROOT", self._encode_mbox_name(mailbox),

File "/srv/modoboa/env/lib/python3.9/site-packages/modoboa_webmail/lib/imaputils.py" in _encode_mbox_name
  415.         return b'"' + folder.encode("imap4-utf-7") + b'"'

During handling of the above exception (unknown encoding: imap4-utf-7), another exception occurred:

File "/srv/modoboa/env/lib/python3.9/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/srv/modoboa/env/lib/python3.9/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "/srv/modoboa/env/lib/python3.9/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/usr/lib/python3.9/contextlib.py" in inner
  79.                 return func(*args, **kwds)

File "/usr/lib/python3.9/contextlib.py" in inner
  79.                 return func(*args, **kwds)

File "/srv/modoboa/env/lib/python3.9/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  21.                 return view_func(request, *args, **kwargs)

File "/srv/modoboa/env/lib/python3.9/site-packages/modoboa/admin/lib.py" in wrapped_f
  37.                 return f(request, *args, **kwargs)

File "/srv/modoboa/env/lib/python3.9/site-packages/modoboa_webmail/views.py" in index
  662.         imapc.getquota(curmbox)

File "/srv/modoboa/env/lib/python3.9/site-packages/modoboa_webmail/lib/imaputils.py" in getquota
  691.             if data is None:
```

**Comment by tonioo** [comment]
Then it might be a bug related to the webmail. Can you open a specific issue?

**Reply by MrGeneration** [reply]
Modoboa imho currently is uninstallable on Debian 11 due to fixed dependency to `feedparser==5.2.1` by modoboa package.
This commit could solve the issue: https://github.com/modoboa/modoboa/commit/2ab3bc3956f12d7311b8e7ad2a5ce285abdd7dba

*Below is dirty tested, however, may break dependencies or even systems. Use with care.*

I've adjusted my `requirements.txt` file:
```
feedparser==6.0.8   # adjusted
sgmllib3k  # new addition, requirement by feedparser ;-/
```

Then I've installed the requirements using `--no-deps` to break explicit package dependencies.
This seems to solve the issue, however, is dirty and might bring issues test may discover before having the feature on a productive environment. A maintenance update to raise dependencies (Django 2.2 is considered insecure / outdated, we're at 4.0 right now) might be a good idea. 

If it doesn't break anything version `1.17.1` should be absolutely fine. :-)
I'm using a very limited modoboa app scope so it might still have issues for others,

I'm not sure how much priority this would have on your list as you're working on the new UI.

**Reply by xword9009** [reply]
Do you happen to know if this is being addressed?

I've just hit the same issue on a clean Debian 11 install using the git installer. 

UI work is cool and all but it would help considerbaly to have a functional server first.

**Reply by tonioo** [reply]
It should be fixed by freshly released 2.0 version
